Improve hooking technique, memory usage and performance

Arcitec opened the pull request:
- Uses an improved method of detecting SuperWoW, and warns the user if it's missing or severely outdated.

- Improved performance and reduced memory usage of the hooks, by only creating the hook functions once at startup, rather than creating new, anonymous functions multiple times every time you right-click anything.

- Keeps track of the current hook-state, so that it doesn't waste time rewriting global variables if the hook is already active/inactive.

- Improved the hooking technique by actually hooking the "set raid icon" API, rather than a pointless wrapper that isn't universally used. This improved hooking technique means that it works with Blizzard's default UI features *and* with 3rd party addons too. It also means that the hooking algorithm is simplified and won't have to recreate Blizzard's wrapper code anymore.

- Documented all important functions to explain how the hook works.

- Added .editorconfig to set the indentation style.

Arcitec committed:
Improve hooking technique, memory usage and performance

- Uses an improved method of detecting SuperWoW, and warns the user if it's missing or severely outdated.

- Improved performance and reduced memory usage of the hooks, by only creating the hook functions once at startup, rather than creating new, anonymous functions multiple times every time you right-click anything.

- Keeps track of the current hook-state, so that it doesn't waste time rewriting global variables if the hook is already active/inactive.

- Improved the hooking technique by actually hooking the "set raid icon" API, rather than a pointless wrapper that isn't universally used. This improved hooking technique means that it works with Blizzard's default UI features *and* with 3rd party addons too. It also means that the hooking algorithm is simplified and won't have to recreate Blizzard's wrapper code anymore.

- Documented all important functions to explain how the hook works.

- Added .editorconfig to set the indentation style.

Arcitec commented:
Thanks for creating the addon. It's a great idea! :heart:

refaim commented:
Thanks for the kind words and for the patch.

I looked at the delta in the pull request and it turns out that almost the entire addon code was rewritten. I like some of the changes, I would argue with some, and I strongly disagree with some.

I am not ready to merge all the changes, but as soon as I have time to work on the addon, I will carefully study them and manually transfer some of them to the addon code.

Arcitec commented:
@refaim Hi Roman, great to hear from you.

Here is a full walkthrough of *all* of the changes:

1. Replaced the slow function calls to `getglobal()` with a fast, local `_G` table instead.
2. Replaced the SuperWoW check with the correct way to check for it, `SUPERWOW_VERSION`, added in SuperWoW 1.2 (we're on 1.5 now).
3. Added a helpful error message telling people that they are missing or have an outdated version of SuperWoW if it's missing.
4. Instead of checking for the existence of SuperWoW over and over and over and over and over and over and over and over and over and over and over again, we check a _single_ time at startup and abort the addon loading if it's missing.
5. Instead of creating lots of memory bloat/churn and wasting CPU cycles by constantly re-creating the fake `GetNumPartyMembers` and `IsPartyLeader` multiple times per right-click (yes, multiple times; it happens every time the person hovers over the Raid Icons submenu item), we now only create them a single time, at addon load, which means no more garbage collector bloat. Way more efficient.
6. We now use `fakeparty_hooked` to track whether the functions are actually hooked, so that we don't waste CPU cycles rewriting the global `GetNumPartyMembers` and `IsPartyLeader` table entries when they're already hooked. This is literally the *only* objectionable change, but I prefer checking the state instead of wasting CPU rewriting multiple variables when there's no need for it.
7. The true WoW API for setting Raid Icons is named `SetRaidTarget`. This is what most 3rd party addons call internally. They don't call the wrapper `SetRaidTargetIcon`, which is Blizzard's wrapper that internally calls `SetRaidTarget`. The old code was rewriting `SetRaidTargetIcon` with a custom version that constantly checks if SuperWoW is loaded (there's that inefficiency again), and had to replicate all Blizzard logic, and did not cover the scenario where addons call the true function (`SetRaidTarget`). So the entire hook was rewritten to a much cleaner solution: Hook `SetRaidTarget`, take the caller's arguments as-is, and only inject the SuperWoW "self_only" icon flag if the caller didn't provide it. This clean hooking means that it automatically works for Blizzard frames and all addons, and always behaves 100% correctly: Setting local icons if you are solo, otherwise sending normal synced icons.
I am very interested in hearing what parts you disagree with and why.

Arcitec commented:
@refaim Hi Roman, great to hear from you. :smiley_cat: 

Here is a full walkthrough of *all* of the changes:

1. Replaced the slow function calls to `getglobal()` with a fast, local `_G` table instead.
2. Replaced the SuperWoW check with the correct way to check for it, `SUPERWOW_VERSION`, added in SuperWoW 1.2 (we're on 1.5 now).
3. Added a helpful error message telling people that they are missing or have an outdated version of SuperWoW if it's missing.
4. Instead of checking for the existence of SuperWoW over and over and over and over and over and over and over and over and over and over and over again, we check a _single_ time at startup and abort the addon loading if it's missing. Perhaps you didn't see the `return` at the top of the file and wondered why I removed the SuperWoW check everywhere? I removed it as part of that performance improvement.
5. Instead of creating lots of memory bloat/churn and wasting CPU cycles by constantly re-creating the fake `GetNumPartyMembers` and `IsPartyLeader` multiple times per right-click (yes, multiple times; it happens every time the person hovers over the Raid Icons submenu item; an average of 6 newly created, anonymous, garbage-collected memory bloat functions were made every time a person right-clicks), we now instead only create them a single time, at addon load, which means no more garbage collector bloat. Way more efficient.
6. We now use `fakeparty_hooked` to track whether the functions are actually hooked, so that we don't waste CPU cycles rewriting the global `GetNumPartyMembers` and `IsPartyLeader` table entries when they're already hooked. This is literally the *only* objectionable change, but I prefer checking the state instead of wasting CPU rewriting multiple global table function references when there's no need for it. Doing this check is more efficient.
7. Renamed variables from `Blizzard_*` to `Original_*`, because "Blizzard" is not the correct label for those "original functions" that we still need to be able to call. They're the originals, but there's no way to know if they are blizzard's variables, since other addons may already have hooked them (so our "originals" can be part of a longer chain of hooked functions; hence the name prefix change to "original").
8. `IsPlayerInPartyOrRaid()` was slightly bugged. It performed its check via the hooked functions, meaning there were situations where the function would look at the fake numbers and return an incorrect result. It has now been fixed: It now uses the original APIs to look at the true party/raid member counts. And `GetNumRaidMembers` has been turned into a local variable to avoid slow global table lookups.
9. The true WoW API for setting Raid Icons is named `SetRaidTarget`. This is what most 3rd party addons call internally. They don't call the wrapper `SetRaidTargetIcon`, which is Blizzard's wrapper that internally calls `SetRaidTarget`. The old code was rewriting `SetRaidTargetIcon` with a custom version that constantly checks if SuperWoW is loaded (there's that inefficiency again), and had to replicate all Blizzard logic, and did not cover the scenario where addons call the true function (`SetRaidTarget`). So the entire hook was rewritten to a much cleaner solution: Hook `SetRaidTarget`, take the caller's arguments as-is, and only inject the SuperWoW "self_only" icon flag if the caller didn't provide it. This clean hooking patches Blizzard's core API and means that it automatically works for Blizzard frames and all addons, and always behaves 100% correctly: Enabling SuperWoW's local client-only icons if you are solo, or otherwise sending normal synced icons if you are in a party/raid. It's a very, very clean way to patch this SuperWoW feature into the game, which automatically "just works" literally everywhere in all Blizzard code and 3rd-party addon code.
10. Changed Lua's default tab-indentation to 4 spaces (which you prefer), and enforced it via `.editorconfig` so that code-formatting tools won't revert it back to tabs.

I am very interested in hearing what parts you disagree with and why. All I want is to give people, users of your addon, more efficient code. If you can point out the parts you disagree with, we can discuss alternative solutions.

It's still your addon and still your idea of how to implement it via menu-hooking and fake functions, but with improved efficiency for your users benefit. It's also difficult to read things in the diff. It's easier to read the code in the plain format, right here, where the logic is very easy to follow:

https://github.com/Arcitec/SoloRaidTargetIcons/blob/master/main.lua
